Pp 4276 part2

oswaldquek opened the pull request:
## WHAT
_A brief description of the pull request:_

## HOW 
_Steps to test or reproduce:_

mrlumbu commented:
Please can you explain more about the what and the how of what you have done. Please can you also look to see if you can split this PR up, or else give some guidance to a reviewer about how to approach reviewing it

oswaldquek commented on src/test/java/uk/gov/pay/connector/util/DatabaseTestHelper.java:
```diff
@@ -493,31 +495,6 @@ public void addEmailNotification(Long accountId, String templateBody, boolean en
-        );
-    }
-
-    public void addCardType(UUID id, String label, String type, String brand, boolean requires3ds) {
```
We've removed the ability for tests to add card types. If a test needs a card type it gets it from the database via methods `getCardTypeByBrandAndType`, `getMastercardCreditCard` etc.

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/resources/HealthCheckResourceITest.java:
```diff
@@ -1,14 +1,22 @@
+import static com.jayway.restassured.http.ContentType.JSON;
 import static org.hamcrest.Matchers.is;
 
-public class HealthCheckResourceITest extends GatewayAccountResourceTestBase {
```
Health check test should not depend on GatewayAccountResourceTestBase

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountResourceTestBase.java:
```diff
@@ -111,26 +118,14 @@ void assertCorrectCreateResponse(ValidatableResponse response, GatewayAccountEnt
         assertThat(accountCredentials, is(new HashMap<>()));
     }
 
-    DatabaseFixtures.TestCardType createMastercardCreditCardTypeRecord() {
```
We no longer allow creation of card types in tests as they are already added via seed data.

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/resources/GatewayAccountFrontendResourceITest.java:
```diff
@@ -452,79 +461,84 @@ public void updateAcceptedCardTypes_shouldNotUpdateGatewayAccountIfCardTypesFiel
-        DatabaseFixtures.TestCardType mastercardCreditCardTypeRecord = createMastercardCreditCardTypeRecord();
-        DatabaseFixtures.TestCardType visaCreditCardTypeRecord = createVisaCreditCardTypeRecord();
-        DatabaseFixtures.TestCardType visaDebitCardTypeRecord = createVisaDebitCardTypeRecord();
-        DatabaseFixtures.TestAccount accountRecord = createAccountRecord(mastercardCreditCardTypeRecord, visaCreditCardTypeRecord);
```
We no longer create card types. The new way is to get already existed card types and associating them with an account.

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/dao/TokenDaoJpaITest.java:
```diff
@@ -54,7 +58,15 @@ public void persist_shouldInsertAToken() {
+        tokenEntity.setId(nextLong());
+        tokenEntity.setToken(tokenId);
+        tokenEntity.setChargeEntity(chargeEntity);
+        tokenDao.persist(tokenEntity);
```
We encountered an issue with adding things directly in the database (which uses JDBI) and getting the data back from a resource (which uses JPA). Do see the commit message on the issue this caused at https://github.com/alphagov/pay-connector/pull/761/commits/e0841844648cd8363132821fa351e62c7864a07b. The right way to make JPA aware of the new entity is to add it using JPA (i.e. the tokenDao).

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/dao/SupportedTypeDaoJpaITest.java:
```diff
@@ -76,30 +72,23 @@ public void findAllNon3ds_shouldFindCardTypes() {
         List<CardTypeEntity> cardTypes = cardTypeDao.findAll();
 
-        assertThat(cardTypes.size(), is(4));
+        assertThat(cardTypes.size(), is(10));
```
There are 10 card types as per the seed data (https://github.com/alphagov/pay-connector/blob/master/src/main/resources/seed-data/card_types.csv)

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/dao/SupportedTypeDaoJpaITest.java:
```diff
@@ -76,30 +72,23 @@ public void findAllNon3ds_shouldFindCardTypes() {
-
-        CardTypeEntity thirdCardType = cardTypes.get(2);
-        assertThat(thirdCardType.getBrand(), is("visa"));
+        assertThat(cardTypes.size(), is(9));
```
There are 9 non 3ds card as per seed data (https://github.com/alphagov/pay-connector/blob/master/src/main/resources/seed-data/card_types.csv)

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/dao/GatewayAccountDaoITest.java:
```diff
@@ -112,61 +117,59 @@ public void findById_shouldFindGatewayAccount() {
 
     @Test
     public void findById_shouldUpdateAccountCardTypes() {
-        DatabaseFixtures.TestCardType mastercardCreditCardTypeRecord = createMastercardCreditCardTypeRecord();
```
We no longer create card types but get them from the existing seed data (https://github.com/alphagov/pay-connector/blob/master/src/main/resources/seed-data/card_types.csv).

oswaldquek commented on src/test/java/uk/gov/pay/connector/it/dao/GatewayAccountDaoITest.java:
```diff
@@ -35,29 +38,31 @@
 
     @Test
     public void persist_shouldCreateAnAccount() {
-        DatabaseFixtures.TestCardType mastercardCreditCardTypeRecord = createMastercardCreditCardTypeRecord();
```
We no longer create card types but get them from the existing seed data (https://github.com/alphagov/pay-connector/blob/master/src/main/resources/seed-data/card_types.csv), and associate the got card types with an account (see changes on the right hand side).

oswaldquek commented on src/test/java/uk/gov/pay/connector/cardtype/dao/CardTypeDaoTest.java:
```diff
@@ -0,0 +1,67 @@
+
+@RunWith(DropwizardJUnitRunner.class)
+@DropwizardConfig(app = ConnectorApp.class, config = "config/test-it-config.yaml")
+public class CardTypeDaoTest {
```
This test was added to test the CardTypeDao against a real database. A convenient way of "spinning" up the database is to run the test with DropwizardJUnitRunner.

oswaldquek commented on src/main/java/uk/gov/pay/connector/cardtype/model/domain/CardTypeEntity.java:
```diff
@@ -79,4 +80,20 @@ public boolean isRequires3ds() {
     }
+
+    @Override
+    public boolean equals(Object o) {
```
This was added as some tests directly compare CardTypeEntities.

oswaldquek commented on pom.xml:
```diff
@@ -6,7 +6,7 @@
     <properties>
         <dropwizard.version>1.3.5</dropwizard.version>
-        <wiremock.version>1.57</wiremock.version>
+        <wiremock.version>2.19.0</wiremock.version>
```
This upgrade solved some random socket issues in tests.

oswaldquek commented:
@mrlumbu Hopefully the explanation below is helpful. I've added some comments within the PR code itself to help explain what's been done as well.

This PR is pretty big and 3 major changes. Unfortunately we weren't what changes broke what so we ended up having to put everything in one PR. Here is an explanation to help the reviewer understand the 3 major changes.

1. The database is set up and torn down once per class, not once per method. This is clearly seen in the changes to DaoITestBase.java. This greatly speeds up things, but caused many test failures because tests weren't using unique charge ids, external charge ids, etc. Thus the vast majority of changes in this PR will be changing hard coded ids to `nextLong()`, and hard coded external charge ids to `RandomIdGenerator.newId()`.

2. We've moved some tests over to use the DropwizardJUnitTestRunner. These are annotated on the test classes. The test classes often extend a base class however, so you'll find some base classes having

```
    @DropwizardTestContext
    protected TestContext testContext;
    protected DatabaseTestHelper databaseTestHelper;
```

3. Changes to the way we manipulate CardTypes in tests. Standard card types are already added via seed data (https://github.com/alphagov/pay-connector/blob/master/src/main/resources/seed-data/card_types.csv) so we've removed the ability for tests to add card types. If a test needs a card type it gets it from the database via the DatabaseTestHelper. Often the purpose of getting a card type(s) is to associate it with a gateway account.

mrlumbu commented on src/test/java/uk/gov/pay/connector/it/dao/ChargeDaoCardDetailsITest.java:
```diff
@@ -19,6 +19,7 @@
 import java.util.Map;
 import java.util.Optional;
 
+import static org.apache.commons.lang.math.RandomUtils.nextLong;
```
total nitpick, but what's wrong with `java.util.concurrent.ThreadLocalRandom.current().nextLong()`?

oswaldquek replied:
nothing. DatabaseFixtures was already using `org.apache.commons.lang3.RandomUtils.nextLong` and ChargingITestBase `org.apache.commons.lang.math.RandomUtils.nextLong`. we just copied `nextLong` and took the first import.

mrlumbu commented on src/test/java/uk/gov/pay/connector/it/dao/GatewayAccountDaoITest.java:
```diff
@@ -35,29 +38,31 @@
-        DatabaseFixtures.TestCardType visaDebitCardTypeRecord = createVisaDebitCardTypeRecord();
-        createAccountRecord(mastercardCreditCardTypeRecord, visaDebitCardTypeRecord);
+        final CardTypeEntity masterCardCredit = databaseTestHelper.getMastercardCreditCard();
+        final CardTypeEntity viseCardDebit = databaseTestHelper.getVisaCreditCard();
```
another nitpick `viseCardDebit` => `visaCardDebit`

mrlumbu commented on src/test/java/uk/gov/pay/connector/it/dao/GatewayAccountDaoITest.java:
```diff
@@ -75,13 +80,13 @@ public void persist_shouldCreateAnAccount() {
-                        org.hamcrest.Matchers.hasEntry("label", mastercardCreditCardTypeRecord.getLabel()),
-                        org.hamcrest.Matchers.hasEntry("type", mastercardCreditCardTypeRecord.getType().toString()),
-                        org.hamcrest.Matchers.hasEntry("brand", mastercardCreditCardTypeRecord.getBrand())
+                        org.hamcrest.Matchers.hasEntry("label", masterCardCredit.getLabel()),
```
is the fully qualified package name needed here?

heathd commented:
![](https://media.giphy.com/media/3oriNYQX2lC6dfW2Ji/giphy.gif)
